source/common/router/rds_impl.cc

```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
     std::weak_ptr<Http::RouteConfigUpdatedCallback> route_config_updated_cb) {
   auto alias =
       VhdsSubscription::domainNameToAlias(config_update_info_->routeConfigName(), for_domain);
-  factory_context_.dispatcher().post([this, alias, &thread_local_dispatcher,
-                                      route_config_updated_cb]() -> void {
-    subscription_->updateOnDemand(alias);
-    config_update_callbacks_.push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-  });
+  // We use weak pointers here as RdsRouteConfigProviderImpl instance could go away before the
+  // dispatcher had a chance to execute the callback.
+  factory_context_.dispatcher().post(
+      [sub = std::weak_ptr<RdsRouteConfigSubscription>(subscription_),
+       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
+       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
+        auto s = sub.lock();
+        auto cb = config_cbs.lock();
+        if (s && cb) {
+          s->updateOnDemand(alias);
+          cb->push_back({alias, thread_local_dispatcher, route_config_updated_cb});
```

Comment:
if in main thread, management server already sent back response with the alias, do you want to send another on demand request to management server?

Depends on management server implementation, if it only sends a response when there is some update of the project, then this HttpRequest will hang till timeout, while Envoy has the config to proceed.

Reply:
This gets pretty complicated quickly: the only time (that I can see) an instance of `RdsRouteConfigProviderImpl`can be deallocated is when LDS is used and the listener configuration has changed. At this point all connections are going to be drained, trying to update a vhost seems not worth it to me.

Even if we wanted to do that, we'd still need an active subscription (may not be there yet, and no way to find out which one to use) and a place to store the callback to notify the on_demand_update filter (which may not be there any more, as the active stream it belongs to may have been closed).

Reply:
I am describing a scenario where no rds subscription destruction is involved. 

could the management server send required vhost to Envoy before the on-demand request is sent? in that case, the management server may or may not send the requested resource again depends on its implementation, in which case the callback will not be called and the request will only timeout.  this is very implementation specific, so I am not too concerned.